wa/instrumentation/hwmon.py

```diff
@@ -0,0 +1,77 @@
+#    Copyright 2017 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from devlib import HwmonInstrument as _Instrument
+
+from wa import Instrument
+from wa.framework.instrumentation import fast
+
+MOMENTARY_QUANTITIES = ['temperature', 'power', 'voltage', 'current', 'fps']
+CUMULATIVE_QUANTITIES = ['energy', 'tx', 'tx/rx', 'frames']
+
+class HwmonInstrument(Instrument):
+    name = 'hwmon'
+
+    description = """
+    Hardware Monitor (hwmon) is a generic Linux kernel subsystem,
+    providing access to hardware monitoring components like temperature or
+    voltage/current sensors.
+    """
+
+    def initialize(self, context):
+        self.instrument = _Instrument(self.target)
+
+    def setup(self, context):
+        self.instrument.reset()
+
+    @fast
+    def start(self, context):
+        self.before = self.instrument.take_measurement()
+
+    def stop(self, context):
```

Comment:
Should be `@fast` as well?

Reply:
Oooo yes, could this also explain why we've been seeing ftrace stopped before the energy instrumentation?

I was planning to investigate that, I assumed that instrumentation should be `stop`ped in the reverse of the order it was `start`ed in...

Reply:
Yes, for the stages after `run()`, the order should be reversed; the priority decorators will take care of of that. `@fast` applied to `stop()` will mean it will before slower stop methods (i.e. closer to the stage it's wrapping).